### PR TITLE
Fix ISO build

### PR DESCRIPTION
**Pull request by davidcassany**
Fixes rancher/elemental-cli#320

Signed-off-by: David Cassany <dcassany@suse.com>

### CODE REVIEW

File: Dockerfile.iso
Comment (davidcassany) [inline Dockerfile.iso:1]: Looks like we have been using the wrong image for a while 🤦🏽

File: elemental-iso-build
Comment (davidcassany) [inline elemental-iso-build:17]: This should not be needed as the elemental-cli already includes the grub.cfg and no download should be needed. However for now I kept it since it includes some weird logic to set a version in a grubenv, I have no idea why we need this, I'd drop it. However I prefer doing it in a separate PR, this is just to fix the broken build. Cosmetic details on grub setup specific to this repository will come later.

File: elemental-iso-build
Comment (davidcassany) [inline elemental-iso-build:30]: This --privileged is needed to build legacy firmware ISOs.... as we need to call `grub2-mkimage` in a chroot for that. I am wondering to completely remove BIOS support when using the `--bootloader-in-rootfs` flag, since the previous strategy of pulling artifacts from somewhere else is still possible it will not be a feature loss in any case.

File: .github/workflows/build-iso.yaml
Comment (davidcassany) [inline .github/workflows/build-iso.yaml:24]: @Itxaka this logic we done from some specific purpose I can't see? This prevents testing script changes on PRs since it always pulls the script from `main`, it does not execute the script form the PR.

I am wondering if this was done for some specific reason I can't see or it is just a leftover.

**Comment by davidcassany** [comment]
We have the failure because in our Dockerfile.image the base image does not include `syslinux`... it will be a mess to include it there, I prefer waiting the base image actually includes it. Gonna disable the offending tests part... for now 😞

**Review by kkaempf — CHANGES_REQUESTED** [review]
ISO building works now, but it fails to boot with
```
'/boot/kernel.xz' not found
```

File: Dockerfile.iso
Comment (ldevulder) [reply Dockerfile.iso:1]: Yes, that was my comment on a previous PR where you were modifying this file ;-)

File: elemental-iso-build
Comment (Itxaka) [reply elemental-iso-build:17]: The grub.cfg file was in there since the start and was copied into the image always.

The version stuff can be dropped, this was for dev purposes and should have no bearing in the final code. In fact we do the same on OBS with an empty version IIRC.

**Comment by davidcassany** [comment]
> ISO building works now, but it fails to boot with
> 
> ```
> '/boot/kernel.xz' not found
> ```

I'd say this is problem of the docker builder cache which does not detect a change in the copied files (the dockerfile remains the same) and does not rebuild the image. I just did `docker builder prune --all --force` and the iso builder image got recreated again, then it included the updated grub.cfg.

File: elemental-iso-build
Comment (Itxaka) [reply elemental-iso-build:30]: that is kind of bad, should not need --privileged to build an iso :/

Is there any way we can circumvent that?

File: .github/workflows/build-iso.yaml
Comment (Itxaka) [reply .github/workflows/build-iso.yaml:24]: That job is not there to prevent broken scripts getting in, it was just added as a extra to test the validity of the script on main. If you check you will see that it doesn't even checkout the source code, so its like an external user getting the script. 

Please add a different script to check the validity of the modified script instead as it has a different scope.

File: elemental-iso-build
Comment (davidcassany) [reply elemental-iso-build:30]: Yes, I am on it. Building for EFI or BIOS firmware images, but not for both... Gonna fix it a follow up PR though. Only BIOS requires privileges.

Comment (Itxaka) [reply elemental-iso-build:30]: we should probably look into dropping bios shortly blargh

**Comment by kkaempf** [comment]
I'm actually not pulling the Dockerfile.iso from GH but have my local copy (which is identical to the `main branch` one, except that it's based on Teal53)

**Review by Itxaka — CHANGES_REQUESTED** [review]
- missing `--privileged` in makefile for local building
 - missing `--bootloader-in-rootfs`in makefile for local building

**Comment by Itxaka** [comment]
> This is just wrong, sorry. This is a chicken and egg problem... the grub config can't be updated in a PR since the script downloads it from main branch... 🤦🏽 Ok, fixing that now...

This is not really a problem. This is a script meant to be used by external users which dont want to clone the repo or anything. That script should only be considered for releases, as it shouldn't be used in any other place. Basically its an extra for non-technical users.

For testing we have the make iso(using local artifacts) and make proper_iso (using OBS artifacts) which again test different things, make iso is mostly for dev envs and local testing, while proper_iso is for testing with remote OBS produced artifacts.

Testing with `make build && make iso` will fail as the `make build` does not install syslinux in the rootfs. `make proper_iso` should work as long as the OBS artifacts have the syslinux packages in them.

Maybe there is a disconnection here as the changes to grub here are the ones that are propagated to the OBS artifacts, and those are the ones then used by the `make proper_iso` but until we merge those grub changes and OBS produces new artifacts, we cannot use `proper_iso`.

The solution here is to add the syslinux packages into our dev image in github (Dockerfile.image) and test the grub changes with those. 

Would be even better if those packages were added on sle-micro-rancher as we wouldn't need to hack our way around.

Thankfully we dont usually need to add packages  because its a trip lol

File: elemental-iso-build
Comment (Itxaka) [inline elemental-iso-build:17]: this breaks the whole point of this script, which is for external users to run the script with a config file and get an iso with no cloning of the repo.

File: elemental-iso-build
Comment (Itxaka) [inline elemental-iso-build:30]: please dont drop the `--squash-no-compression` as it saves a LOT of time in building isos, especially on github :)

**Comment by davidcassany** [comment]
> This is not really a problem. This is a script meant to be used by external users which dont want to clone the repo or anything. That script should only be considered for releases, as it shouldn't be used in any other place. Basically its an extra for non-technical users.

I am fine with this approach, but then we should remove the script testing form the PR given that it does not make use of the source form the PR itself. Gonna re-add the files download form _main_ branch then.
 
> For testing we have the make iso(using local artifacts) and make proper_iso (using OBS artifacts) which again test different things, make iso is mostly for dev envs and local testing, while proper_iso is for testing with remote OBS produced artifacts.

Ok I see.
 
> Testing with `make build && make iso` will fail as the `make build` does not install syslinux in the rootfs. `make proper_iso` should work as long as the OBS artifacts have the syslinux packages in them.

Yes definitely, this will fail until syslinux is in the rootfs or until we are capable to build pure EFI images.
 
> Maybe there is a disconnection here as the changes to grub here are the ones that are propagated to the OBS artifacts, and those are the ones then used by the `make proper_iso` but until we merge those grub changes and OBS produces new artifacts, we cannot use `proper_iso`.

Not really, `proper_iso` should be already functional

> 
> The solution here is to add the syslinux packages into our dev image in github (Dockerfile.image) and test the grub changes with those.

I will not do that, it will be really messy. There are not sle micro repositories available as far as I know, so the image comes with no repositories enabled. I don't want to start adding third party binaries here or define opensuse repos on a sle-micro image.

> 
> Would be even better if those packages were added on sle-micro-rancher as we wouldn't need to hack our way around.

definitely

**Comment by kkaempf** [comment]
`syslinux` is part of the OBS node-image/5.3 and has been requested for SLE-Micro-for-Rancher

**Comment by Itxaka** [comment]
> I am fine with this approach, but then we should remove the script testing form the PR given that it does not make use of the source form the PR itself. Gonna re-add the files download form main branch then.

Yes. That was actually a mvp script. It needs to be fully reworked as we may want to include it on the releases (so people can download just that) and that would need to have a fixed tag that we would need to generate. So yeah, it was there becasue it was requested but the bare minimum was done with it. Didnt even noticed that it runs on PR lol, probably a leftover of my testing while adding it :D

> I will not do that, it will be really messy. There are not sle micro repositories available as far as I know, so the image comes with no repositories enabled. I don't want to start adding third party binaries here or define opensuse repos on a sle-micro image.

Well we kind of already do that, dont we? For our local testing, as we are building the images ourselves, we rely on luet to provide some needed packages on top of the OS (we could probably drop all of them LMAO). Talking about luet packages here, not usual RPM packages.

If we dont do that, dev is stopped until the package trickles down into SLE-Micro-for-Rancher as github uses that as base. No e2e tests, no local dev of anything that needs an iso, as our local changes do not get into the iso for testing. No way of testing anything or developing as we wont be able to create an iso with the current setup.

We can use the remote iso but that doesnt include any local changes done here. :shrug:

File: elemental-iso-build
Comment (Itxaka) [inline elemental-iso-build:30]: I wonder why it needs the full path to the overlay now...is it due to the chroot?

**Comment by Itxaka** [comment]
Im still baffled on where the grub.cfg file for the live iso comes from on this patch? I dont see where it can come from, there is no grub.cfg in the framework and the one specifically to build the iso is gone. And I dont see it being generated as part of the build-iso or anything?

what Im I missing?

**Comment by Itxaka** [comment]
> Im still baffled on where the grub.cfg file for the live iso comes from on this patch? I dont see where it can come from, there is no grub.cfg in the framework and the one specifically to build the iso is gone. And I dont see it being generated as part of the build-iso or anything?

Ah finally found it, was looking in the wrong place :D

File: elemental-iso-build
Comment (davidcassany) [reply elemental-iso-build:30]: Yes I am also so puzzled because of this... 😞

File: elemental-iso-build
Comment (davidcassany) [reply elemental-iso-build:30]: This is part of the manifest.yaml now. I placed there the non-dynamic flags.

**Comment by davidcassany** [comment]
@kkaempf @Itxaka ok the current patch should let you run `make proper_iso` without issues. `make build && make iso` also works since I finally opted to include syslinux from Leap in Dockerfile.image (god forgive... ). At least this way the ISO creation should be unblocked.

`elemental-iso-build` script fails because it will attempt to download files from main branch that are not yet in main branch.

File: .github/workflows/build-iso.yaml
Comment (davidcassany) [reply .github/workflows/build-iso.yaml]: Finally kept as it was

File: elemental-iso-build
Comment (davidcassany) [reply elemental-iso-build]: reverted

**Review by ldevulder — APPROVED** [review]
ISO compilation is fixed, I think that's enough for that PR :-) Let's fix the other stuffs in new PR.

**Review by Itxaka — APPROVED** [review]

File: Dockerfile.image
Comment (ldevulder) [reply Dockerfile.image:49]: For a workaround what is the benefit? Anyway if it's a quick change that's fine.

I'm saying this because this issue is blocking the CI and also manual testing with ISO, so if we could have a fix it could be great :-)

Comment (Itxaka) [reply Dockerfile.image:49]: mainly just wondering why he choose this over adding our toolkit package, knowing that it was used for the isos before(I think?)

Maybe our package is broken or not valid for this workaround. 

Note that this is just a question, no blocking the PR, I actually approved it ;)

Comment (Itxaka) [reply Dockerfile.image:49]: oh shit, we have the resolve conversations before merging enabled so that is why this wasnt merged automatically. Sorry about that, didnt thougth we had that enabled anywhere as its a pita. Resolving this comment so it gets merged

**Comment by Itxaka** [comment]
we also had a pending changes requested. I thougth those got cleared on new commits :/

**Comment by juadk** [comment]
I have an issue with the new version of the script:
```
vagrant@k3s-opensusetw:~> ./elemental-iso-build my-registration.yaml
Using custom registration from my-registration.yaml
curl: (3) URL using bad/illegal format or missing URL
```

**Comment by kkaempf** [comment]
Confirmed fixed. Running `elemental-iso-build` produces a working ISO again. 
Thanks !

**Pull request merged**
